STL_Extension/include/CGAL/Small_unordered_set.h

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2020 GeometryFactory Sarl (France)
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org)
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Simon Giraudot
+
+#ifndef CGAL_SMALL_UNORDERED_SET_H
+#define CGAL_SMALL_UNORDERED_SET_H
+
+#include <array>
+#include <unordered_set>
+
+namespace CGAL
+{
+
+/*
+  This is a very rudimentary structure. It is far from being a full
+  "small unordered set", but it is a starting point.
+
+  For the moment, its only feature is the insertion of elements +
+  check that they were indeed inserted, while avoiding instantiating a
+  `std::unordered_set` as long as the number of elements remain small.
+
+  In practice:
+
+  - if the number of elements in the set is lower than MaxSize,
+    elements are just appended to a `std::array<Key, MaxSize>` and the
+    unicity test is done element by element, in linear time
+
+  - when the number of elements exceed MaxSize, a
+    `std::unordered_set<Key>` is instanciated, all the elements of the
+    array are inserted in it and from that point the container behaves
+    like a `std::unordered_set`
+
+  For this structure to be a true "small unordered set", special
+  iterators should be created to make the switch from array to
+  unordered set completely transparent to the user, and all other
+  usual member functions should be introduced. So far, this is not
+  needed and thus not done.
+*/
+template<typename Key, std::size_t MaxSize>
+class Small_unordered_set
+{
+  using Array = std::array<Key, MaxSize>;
+  using Set   = std::unordered_set<Key>;
+
+  Array m_array;
+  std::shared_ptr<Set> m_set;
```

Comment:
```
constexpr static std::size_t MaxSize = sizeof(Set) / sizeof(Key);
// Are you really needed MaxSize? May be this?
union {
    Set set;// typical sizeof(std::unordered_set) is 56~
    Array array;
} impl;
uint8_t size;

Small_unordered_set() {
    size = 0;
    impl.array = Array{};
}

bool insert(const Key &key) {
    if (size <= MaxSize) {
        for (const Key &a : impl.array) {
            if (a == key) {
                return false;
            }
        }
        if (size < MaxSize) {
            impl.array[size++] = key;
            return true;
        } else {
            Array temp = std::move(impl.array);
            impl.set = Set{};
            impl.set.reserve(++size);
            impl.set.insert(temp.begin(), temp.end());
        }
    }
    return impl.array.set.insert(key).second;
}
```

Reply:
Even if you decide to use the pointer use unique_ptr instead shared

Reply:
I think you missed the point of this set: it is not to reduce memory usage but to increase speed. Max size should be tuned to minimize computation times (not to match the "weight" of an unordered set). Finding the value in an array is done linearly and we certainly don't want to do it for large arrays.

Using a union just for the sake of saving one pointer seems unnecessary to me. Also using unions of non-simple types also implies calling destructors by hand and overall makes the code less readable and more error-prone.

Shared pointer is used because this structure can be copied.

Reply:
I understood correctly, you just use MaxSize == 8, and I thought that maybe it is worth just adjusting them

Union is standard practice for sso

If you want easy cope this container, you should use shared_ptr<Container> in curve class, because otherwise the semantics of copying are not obvious, besides, why would you need to copy the array every time